### PR TITLE
Vaccines List View

### PR DESCRIPTION
**Pull request by juliatan**
# Description

NOTE: Still debugging/fixing "back" issue on add/edit vaccine which goes to tests list vs vaccine list.

Updates the vaccine screens to the new design, which allows many vaccines (actually doses). Has A/B feature flag.

Backend PR (run locally if testing) https://github.com/zoe/covid-app/pull/542 

Notion/Figma
https://www.notion.so/joinzoe/Vaccine-Logging-List-View-efae05e25ccd4f439070f2f389e5ad96
https://www.figma.com/file/7TEcX7f6rw3Z0WYR4TpX3R/ZOE-COVID-Study-(App)---MASTER?node-id=12017%3A138203

## On what platform have you tested the change?

- [ ] Android - Emulator
- [ ] Android - Physical device
- [x] iOS - Emulator
- [ ] iOS - Physical device

## Screenshots

[comment]: <> (Please attach screenshots showing the change, if relevant)

## Checklist (for submission)

- [ ] Analytics/tracking events have been added (if relevant)
- [ ] Added new screen(s) to `src/features/screens.ts` and new Redux slice(s) to `mocks/mockedInitialState.ts` (if relevant)

## Checklist (for reviewers)

- [ ] Checked against Figma designs (if relevant)
- [ ] Checked that data has been successfully saved to the backend (if relevant)

## Out of scope and potential follow-up

[comment]: <> (Are there any related changes that you plan to do in a follow-up PR or voluntarily excluded from the scope?)

### CODE REVIEW

File: src/core/state/vaccines/slice/index.ts [hunk 1]
```diff
@@ -7,13 +7,14 @@ export const initialStateVaccine: IVaccineState = {
...
 };
 
-export const fetchVaccines = createAsyncThunk('vaccines/fetchVaccines', async (patientId: string): Promise<
-  TVaccineRequest[]
-> => {
-  const response = await vaccineService.listVaccines();
-  const patientVaccines = response.filter((vaccine) => vaccine.patient === patientId);
-  return patientVaccines;
-});
+export const fetchVaccines = createAsyncThunk<unknown, string>(
```
Comment (juliatan) [inline src/core/state/vaccines/slice/index.ts:10]: I'm not sure why the decision was made to call the API endpoint in the Redux slice, considering the list of vaccines are not then added to the state. Do you guys think I should leave here or move call to the component? 🤔

Comment (bionara) [reply src/core/state/vaccines/slice/index.ts:10]: I think move to the component then it's one less async thunk to worry about in the codebase!

**Review by bionara — DISMISSED** [review]
Looks good so far!

File: src/features/vaccines/components/VaccineDoseRow.tsx [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+};
+
+export const VaccineDoseRow: React.FC<TProps> = ({ dose, testID, vaccineRecord, index }) => {
+  const formatVaccineDate = (dose: TDose) => {
+    return moment(dose.date_taken_specific).format('MMM D, Y');
+  };
+
+  const brand = (dose: TDose) => {
+    // TODO: What if brand is not set?
+    return dose.brand ? vaccineBrandDisplayName[dose.brand] : 'Unknown';
```
Comment (aikewoody) [inline src/features/vaccines/components/VaccineDoseRow.tsx:23]: Should this be translated?

File: src/features/vaccines/fields/VaccineDoseQuestion.tsx [hunk 3]
```diff
@@ -82,13 +91,40 @@ export const VaccineDoseQuestion: IVaccineDoseQuestion<IProps, IVaccineDoseData>
...
+      </View>
+    );
+  };
+
+  const renderPickerUpdated = () => {
+    const dateField: Date | undefined = formikProps.values.doseDate;
+    let minDate: Date | undefined;
+
+    if (isGBCountry()) {
+      minDate = new Date('2020-12-08');
```
Comment (aikewoody) [inline src/features/vaccines/fields/VaccineDoseQuestion.tsx:111]: Why specifically this date?

File: src/features/vaccines/fields/VaccineDoseQuestion.tsx [hunk 4]
```diff
@@ -82,13 +91,40 @@ export const VaccineDoseQuestion: IVaccineDoseQuestion<IProps, IVaccineDoseData>
...
+    }
+    if (isUSCountry()) {
+      minDate = new Date('2020-12-11');
```
Comment (aikewoody) [inline src/features/vaccines/fields/VaccineDoseQuestion.tsx:114]: Why specifically these dates?

File: src/features/vaccines/__tests__/VaccineListScreenUpdated.test.tsx [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    const elementBase = <MockedNavigator Component={VaccineListScreenUpdated} />;
+    const elementWithRedux = <ReduxProvider store={store}>{elementBase}</ReduxProvider>;
+    const element = <ThemeProvider theme={theme}>{elementWithRedux}</ThemeProvider>;
+    const instance = renderer.create(element).root;
+    await flushPromises();
+    await act(async () => {
+      expect(instance.findAllByType(VaccineDoseRow).length).toBe(0);
+    });
+  });
+});
```
Comment (aikewoody) [inline src/features/vaccines/__tests__/VaccineListScreenUpdated.test.tsx:87]: Loving all the tests you're writing!

File: src/features/vaccines/components/VaccineDoseRow.tsx (hunk 2 above)
Comment (bionara) [reply src/features/vaccines/components/VaccineDoseRow.tsx:23]: I think this is more to stop a crash on key error, ie return a string: in theory, this should never happen as all doses need brand.

File: src/features/vaccines/fields/VaccineDoseQuestion.tsx (hunk 4 above)
Comment (bionara) [reply src/features/vaccines/fields/VaccineDoseQuestion.tsx:114]: these are the "launch dates" for vaccines in those countries: minDate stops vaccines being entered before those dates

File: src/NavigatorService.ts [hunk 6]
```diff
@@ -32,6 +32,10 @@ function goBack() {
   navigation?.goBack();
 }
 
+function push<RouteName extends TScreenName>(routeName: RouteName, params?: TScreenParamList[RouteName]) {
```
Comment (bionara) [inline src/NavigatorService.ts:35]: @Aike added the navigation push method to our navigator service. I needed it to be used to “fake” an entry on nav stack as the feature toggle screen for vaccines serves multiple views.

File: src/core/assessment/AssessmentCoordinator.ts [hunk 7]
```diff
@@ -171,12 +182,15 @@ export class AssessmentCoordinator extends Coordinator {
     });
   };
 
-  goToAddEditVaccine = (vaccine?: TVaccineRequest) => {
+  goToAddEditVaccine = (vaccine?: TVaccineRequest, index?: number) => {
     if (vaccine) {
       this.assessmentData.vaccineData = vaccine;
     }
-    NavigatorService.navigate('AboutYourVaccine', {
+    // FeatureToggle: Use push to add "another" state to nav stack so that its back returns to "itself" (defaults to LIST view)
```
Comment (bionara) [inline src/core/assessment/AssessmentCoordinator.ts:189]: @Aike using the push here btw

File: src/features/vaccines/components/VaccineDoseRow.tsx (hunk 2 above)
Comment (aikewoody) [reply src/features/vaccines/components/VaccineDoseRow.tsx:23]: Got it, then the `TODO` can be removed, right?

File: src/features/vaccines/fields/VaccineDoseQuestion.tsx (hunk 4 above)
Comment (aikewoody) [reply src/features/vaccines/fields/VaccineDoseQuestion.tsx:114]: Ah interesting, let's insert this knowledge in the code. Could be my name it e.g. `vaccinationStartDateUS` or add some comments above it

File: src/core/vaccine/VaccineApiClient.ts [hunk 8]
```diff
@@ -47,6 +47,7 @@ export class VaccineApiClient implements IVaccineRemoteClient {
     return apiClient.get<TVaccineRequest[]>(`/vaccines/`);
   }
 
+  // Depricated as the "new" Vaccine system treats vaccine as a "Vaccination Record", and uses PATCH to "delete" doses
```
Comment (aikewoody) [inline src/core/vaccine/VaccineApiClient.ts:50]: deprecated :)

File: src/core/vaccine/VaccineService.ts [hunk 9]
```diff
@@ -39,6 +39,7 @@ export class VaccineService implements IVaccineService {
     return vaccineApiClient.listVaccines();
   }
 
+  // Depricated as the "new" Vaccine system treats vaccine as a "Vaccination Record", and uses PATCH to "delete" doses
```
Comment (aikewoody) [inline src/core/vaccine/VaccineService.ts:42]: deprecated :)

File: src/core/vaccine/dto/VaccineRequest.ts [hunk 10]
```diff
@@ -4,23 +4,26 @@ export type TVaccineRequest = {
...
   version: string; // document/schema version
   patient: string; //	Patient id
-
-  vaccine_type: EVaccineTypes;
-  brand: EVaccineBrands | null;
-  placebo: EPlaceboStatus | null;
   doses: Partial<TDose>[];
-  description: string; // eg 'mRNA'
+
+  // These are depricated fields; this data now lives on the Dose itself
```
Comment (aikewoody) [inline src/core/vaccine/dto/VaccineRequest.ts:9]: deprecated :)

File: src/features/vaccines/AboutYourVaccineScreenOld.tsx [hunk 11]
```diff
@@ -195,7 +195,7 @@ export function AboutYourVaccineScreen({ route }: TProps) {
       <Formik
         validateOnChange
         validateOnMount
-        initialValues={{ ...buildInitialValues(assessmentData?.vaccineData) }}
+        initialValues={{}}
```
Comment (aikewoody) [inline src/features/vaccines/AboutYourVaccineScreenOld.tsx:198]: Can this be omitted completely?

File: src/features/vaccines/AboutYourVaccineScreenUpdated.tsx [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+    };
+  };
+
+  const navigation = useNavigation();
+
+  const renderDeleteOrBack = () => {
+    return doseBeingEdited ?
+      (
+        <TouchableOpacity onPress={() => promptDelete()}>
+          <ClickableText onPress={() => promptDelete()} style={styles.clickableText}>
```
Comment (aikewoody) [inline src/features/vaccines/AboutYourVaccineScreenUpdated.tsx:127]: Can both be shortened to `onPress={promptDelete}`

File: src/features/vaccines/VaccineListFeatureToggleScreen.tsx [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+  if (route.params?.viewName === 'ADD_EDIT_DELETE') {
+    return startupInfo?.show_new_vaccines_ui ? (
+      <AboutYourVaccineScreenUpdated route={route} />
+    ) : (
+      <AboutYourVaccineScreenOld route={route} />
+    );
+  }
+
+  return null;
+};
```
Comment (aikewoody) [inline src/features/vaccines/VaccineListFeatureToggleScreen.tsx:36]: Inventive idea! 👍

File: src/features/vaccines/VaccineListFeatureToggleScreen.tsx [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+export const VaccineListFeatureToggleScreen: React.FC<TProps> = ({ route }) => {
+  const startupInfo = useSelector<TRootState, TStartupInfo | undefined>((state) => state.content.startupInfo);
+
+  if (route.params?.viewName === 'LIST') {
+    return startupInfo?.show_new_vaccines_ui ? (
+      <VaccineListScreenUpdated route={route} />
+    ) : (
+      <VaccineListScreenOld route />
```
Comment (aikewoody) [inline src/features/vaccines/VaccineListFeatureToggleScreen.tsx:23]: I guess this must be `route={route}`?

File: src/features/vaccines/VaccineListFeatureToggleScreen.tsx [hunk 15]
```diff
@@ -0,0 +1,36 @@
+import { TRootState } from '@covid/core/state/root';
+import { TStartupInfo } from '@covid/core/user/dto/UserAPIContracts';
+import { TScreenParamList } from '@covid/features/ScreenParamList';
+import { RouteProp } from '@react-navigation/native';
+import * as React from 'react';
+import { useSelector } from 'react-redux';
+import { VaccineListScreenOld } from '.';
+import { AboutYourVaccineScreenOld } from './AboutYourVaccineScreenOld';
+import { AboutYourVaccineScreenUpdated } from './AboutYourVaccineScreenUpdated';
+import { VaccineListScreenUpdated } from './VaccineListScreenUpdated';
```
Comment (aikewoody) [inline src/features/vaccines/VaccineListFeatureToggleScreen.tsx:10]: Let's keep using absolute imports :)

File: src/features/vaccines/AboutYourVaccineScreenUpdated.tsx [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+              <View style={styles.footerWrapper}>
+                {!!Object.keys(props.errors).length && props.submitCount > 0 ? (
+                  <ValidationError error={i18n.t('validation-error-text')} style={styling.marginBottom} />
+                ) : null}
+
+                <BrandedButton enabled={props.isValid} onPress={props.handleSubmit} testID="button-submit">
+                  {i18n.t('vaccines.your-vaccine.save-dose')}
+                </BrandedButton>
+
+                {renderDeleteOrBack()}
```
Comment (aikewoody) [inline src/features/vaccines/AboutYourVaccineScreenUpdated.tsx:170]: Just curious about this, when do you decide to extract JSX render logic into its own function?

**Review by aikewoody — CHANGES_REQUESTED** [review]
Massive PR mate and great work! Overall looks good but would like to see some comments I placed resolved, besides that you have my approval 👍

File: src/features/vaccines/AboutYourVaccineScreenUpdated.tsx (hunk 16 above)
Comment (bionara) [reply src/features/vaccines/AboutYourVaccineScreenUpdated.tsx:170]: Right away. Makes the render method a lot easier to read and work out the "chunks" of "logic"

**Review by aikewoody — APPROVED** [review]
Let's merge this baby in! (or out haha)

**Pull request merged**